rofl-scheduler/src/state.rs

```diff
@@ -0,0 +1,99 @@
+use std::{collections::BTreeMap, time::Instant};
+
+use base64::{prelude::BASE64_STANDARD, Engine};
+use oasis_runtime_sdk::types::address::Address;
+use oasis_runtime_sdk_rofl_market::types::{Deployment, Instance};
+
+use crate::types::{Permissions, METADATA_KEY_PERMISSIONS};
+
+/// Instance state.
+#[derive(Default, Debug, Clone)]
+pub struct InstanceState {
+    /// Address of the instance administrator.
+    pub admin: Address,
+    /// Permissions.
+    pub permissions: Permissions,
+    /// Last deployment.
+    pub last_deployment: Option<Deployment>,
+    /// Last error message corresponding to deploying `last_deployment`.
+    pub last_error: Option<String>,
+    // Whether to ignore instance start until the given time elapses.
+    pub ignore_start_until: Option<Instant>,
+    /// Backoff associated with ignoring instance start.
+    pub ignore_start_backoff: Option<backoff::ExponentialBackoff>,
+}
+
+impl InstanceState {
+    /// Updates instance state from the given instance and deployment.
+    pub fn update_from(&mut self, instance: &Instance, deployment: Option<&Deployment>) {
```

Comment:
Could we split this into two methods, `update_admin` and `update_permissions`? Or rename the `update_from` to something less general, since the method name does not clearly indicate what is getting updated.